Configure Decidim Bulletin Board

leio10 opened the pull request:
#### :tophat: What? Why?
This PR implements the first step needed to configure a connection to a Bulletin Board from Decidim. It generates an identification pair of keys to be able to sign all the requests sent to the Bulletin Board. 

#### :pushpin: Related Issues
- Related to #6288
- Fixes #6286

#### :clipboard: Subtasks
- [ ] Add `CHANGELOG` upgrade notes, if required
- [ ] If there's a new public field, add it to GraphQL API
- [x] Add documentation regarding the feature 
- [x] Add tests

### :camera: Screenshots (optional)
* Rake task call
![task private key](https://user-images.githubusercontent.com/453545/90484669-ca7e9b80-e136-11ea-8c23-a894fdb53a0e.png)
![task public key](https://user-images.githubusercontent.com/453545/90484675-cc485f00-e136-11ea-8470-1f7c30be8b5f.png)
* Bulletin Board not set warning
![no bulletin board warning](https://user-images.githubusercontent.com/453545/90484609-b5a20800-e136-11ea-9ec1-957c817eddc1.png)

leio10 commented:
@decidim/product I've some concerns about this implementation:
- some Decidim instances, as Decidim Barcelona, are deployed to Heroku. This service uses ephemeral storage, meaning that all created files will be lost on each deploy. So, creating the files with the task seems a bit problematic, as we can't upload the private key to the repository nor create it directly on the server on those cases. That's why I've decided to print the keys on the screen and explain different options in the documentation.
- as @tramuntanal points [here](https://github.com/decidim/decidim-bulletin-board/pull/1#discussion_r466183550), secure public keys are multiline and too lengthy to be used as parameters for rake tasks, even when it's possible. I can think of three options here: 
  - keep it as it is now, using quotes to call the add_authority task:

    ```sh
    $ export PK="-----BEGIN PUBLIC KEY-----
    MIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA32d9tY44lfV6OdanXQ4B
    102oYFEYMouZUDHquD5yTNZZkTYJ71u9VCHtVf83rW+YUj8NZFtoCeo/g9TUwmKf
    nUl362MeNlrpbTUU2yD3rJjt5JDme4+gtFPjTneubJEOwwHBKJpDNqZB+HmxNqbz
    3nn4EoBqv8tSKFon8Mk1Nsj9aPW3jpHnVCjWUJ8wtB321TgYW2GMDv8aOkOZ/M7p
    z4pFk49Skp24knITeeKdzB2VvcRhEdAf8Lp5Jqtoh9za40Ky0AvPx/Ueh42xH9t+
    P9hfQFX5joSlCKMAPe1Xyji2gxiwWTBu60mOL7PbThiDxrOZxSVhhN21A4IrDfIN
    T2PJCDtYZUyqu6xYISPg6i3teXZrPWT4wWVcJ/dPEXJwSomuFpK7AZ07uhwSwLPw
    pZc/yUxDq3xNfyVCRfu9UoCvrKwKMK7BrXPzM6ljtD9KzjhqbXb1q+JQXDp6uxIH
    FV8TXJSZS7o2tU9G0AoXiKRIOS1t6IG19YhC6Yqn5raPjRlVO//suJS+WAV+j3tU
    SHFOHWo6HYXcS6WRHItkudXjwnyIhW06Zs+KuMKXQQmEkytcFi562pwvUH7kvAZ2
    8G0TR5G5wgRbrGQzM4WyMuemWbcFpG6RqwboffHNCkxSetDHY9WOF3EQ0pKqzaJB
    XbnNp1nB7AHiy2yxCQznpUMCAwEAAQ==
    -----END PUBLIC KEY-----"

    $ bin/rails "client:add_authority[Authority name, $PK]"
    ```

   - dump the public key to a file and use it to send it to the Bulletin Board administrator, as @tramuntanal suggested
   - add it to the graphql and change the Bulletin Board task to fetch all the Decidim instance from the graphql response, only needing the graphql endpoint, something like this:
     
     ![image](https://user-images.githubusercontent.com/453545/90503194-500e4580-e14f-11ea-8516-7956b5d0337a.png)

     I like this solution, as it is pretty straightforward for administrators, but I'm not very convinced about adding a graphql client to the BB and dependency with Decidim, that currently we don't have (but it may be needed in the future).

- As shown in the PR screenshots part, I've added a warning message to warn the administrators that the Elections module can't be used without a configured BB. This alert will appear on every visit to the elections list page until the identification keys are generated and the BB settings are set. I don't know if think that it's ok or if this could be too annoying for administrators.

That's all. As you can see, in this PR I've decided to implement the simplest approach to be able to iterate over it until we find the best solution for each concern.

tramuntanal commented:
I have to contradict myself but taking into account environments like Heroku that have ephemeral file systems it may be best to keep with the current implementation and pass the public key as a task parameter

andreslucena commented:
@leio10 I've reformated your answer (by giving a numbered alphabetical list), so it's easier to follow/discuss.

What I was thinking initially regarding the key with the add_authority task is to have something like the one you have in ssh (ie `.ssh/id_rsa.pub`), that'd be less error-prone for commands IMHO. 

Also in the future, after we have the first working implementation, we should make a step by step tutorial with examples of format, as this could be an issue. For instance in your `c.` example there are newline escape sequences (`\n`)

tramuntanal commented:
@leio10 what's the state of this PR?

andreslucena commented on decidim-elections/config/locales/en.yml:
```diff
@@ -88,6 +88,7 @@ en:
             title: Edit election
             update: Update election
           index:
+            no_bulletin_board: There is no Bulletin Board server configured, which is needed to use this module. This task should be done by the System Administrator.
```
Can we add a link to https://github.com/decidim/decidim-bulletin-board :pray: ?

andreslucena commented on docs/services/elections_bulletin_board.md:
```diff
@@ -0,0 +1,36 @@
+# Elections Bulletin Board
+
```
Please add a line like: "This is a work in progress and is not fully working yet."

leio10 commented:
@andreslucena I've addressed your comments. Regarding the link you asked to add in the warning message, it was shown inside a callout, that wasn't prepared to show links. So, I've added some styling for that. You can check it [here](https://decidim-staging-pr-165.herokuapp.com/admin/participatory_processes/eos-fugit/components/10/manage/?locale=es).

andreslucena reviewed (approved):
:rocket:

leio10 commented:
@decidim/core this was approved by Product and is ready to be reviewed

tramuntanal commented on decidim-elections/lib/decidim/elections/bulletin_board_client.rb:
```diff
@@ -0,0 +1,35 @@
+      attr_reader :identification_private_key, :server, :api_key, :private_key
+
+      def identification_private_key_content
+        @identification_private_key_content ||= if identification_private_key.starts_with?("-----")
+                                                  identification_private_key
+                                                else
+                                                  File.read(Rails.application.root.join(identification_private_key))
+                                                end
```
Good, two ways to set the private key. I see it is also documented :clap:

tramuntanal reviewed (approved):
Great work @leio10 !
